net: wifi: Introduce Wi-Fi network manager's support

krish2718 opened the pull request:
This introduces support for Wi-Fi network managers in Zephyr. The
    motivation is for the Wi-Fi management layer to work with both
    Network managers and offloaded Wi-Fi drivers. The device driver
    decides which one to use.
    
    network manager : Apps -> Wi-Fi Mgmt -> Network Manager -> Wi-Fi
    interface
    
    offloaded       : Apps -> Wi-Fi Mgmt -> Wi-Fi offloaded interface
    
    Support for multiple network managers has been added, each device can
    choose its own network manager and there can be mix and match:
    
      wlan0 - Offloaded
      wlan1 - Network manager 1
      wlan2 - Network manager 2

krish2718 committed:
net: wifi: Move Wi-Fi ops to a separate struct

Decouple interface and Wi-Fi APIs, Wi-Fi APIs are common independent of
Wi-Fi offload or implemented natively (This is preparation for
introducing Native Wi-Fi).

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
net: wifi: Introduce Wi-Fi network managers

This introduces support for Wi-Fi network managers in Zephyr. The
motivation is for the Wi-Fi management layer to work with both
Network managers and offloaded Wi-Fi drivers. The device driver
decides which one to use.

network manager : Apps -> Wi-Fi Mgmt -> Network Manager -> Wi-Fi
interface

offloaded       : Apps -> Wi-Fi Mgmt -> Wi-Fi offloaded interface

Support for multiple network managers has been added, each device can
choose its own network manager and there can be mix and match:

  wlan0 - Offloaded
  wlan1 - Network manager 1
  wlan2 - Network manager 2

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
net: wifi: Add support to handle Wi-Fi NM's in Wi-Fi management

Check if a network interface is managed by a network manager before
falling back to offload API.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

rlubos commented on include/zephyr/net/net_if.h:
```diff
@@ -2535,6 +2535,13 @@ bool net_if_is_suspended(struct net_if *iface);
+ *
+ * @return Pointer to network interface, NULL if not found.
+ */
+struct net_if *net_if_get_first_wifi(void);
```
I think commits:
* `net: wifi: Add support to get and set Wi-Fi as default interface`
*  `net: wifi: Use only Wi-Fi interfaces`

could be a part of #58869 - they don't seem related to the rest of `wifi_mgmt` changes in here, and are a nice supplement to changes in #58869, as they actually show the reason for those changes.

rlubos commented on include/zephyr/net/ethernet.h:
```diff
@@ -352,8 +352,11 @@ enum ethernet_if_types {
 	L2_ETH_IF_TYPE_WIFI,
+
+	/** Native IEEE 802.11 Wi-Fi */
+	L2_ETH_IF_TYPE_WIFI_NATIVE,
```
After looking at the logic in commit `net: wifi: Add support to handle Wi-Fi native to Wi-Fi management` - is this separate `WIFI_NATIVE` type even needed? It doesn't seem to rely on the interface type, but rather have a nice fallback mechanism - if there's no "shared" implementation found, it falls back to offloaded one.

krish2718 replied:
Well, if we have a `type` to convey, then this `enum` is needed to convey. It can also be used as a check for Native stack to ensure given interface is `native`.

krish2718 replied:
I have moved the two commits over to the other PR, and I will remove the WIFI_NATIVE altogether.

rlubos commented on subsys/net/l2/wifi/Kconfig:
```diff
@@ -27,3 +27,20 @@ config WIFI_MGMT_RAW_SCAN_RESULTS_ONLY
 
 endif # WIFI_MGMT_RAW_SCAN_RESULTS
+
+config WIFI_NATIVE
```
IMO (but it's not only my opinion) the "wifi_native" nomenclature is a bit confusing. Especially "native Wi-Fi stack"- it suggests that we have a complete Wi-FI stack (including MAC) for Zephyr. In reality though, if I understood the intent of this PR correctly, it's more about providing a framework to register a Wi-Fi supplicant instance, that can be shared among multiple interfaces.

krish2718 replied:
Agreed, even I was thinking the same after publishing this PR, but thought will wait for others to chime in. Should we just call this `supplicant` support? but it might be mistaken with `wpa_supplicant`.

rlubos commented on include/zephyr/net/wifi_native.h:
```diff
@@ -0,0 +1,119 @@
+struct wifi_native_mgd_iface {
+	/** Interface */
+	struct net_if *iface;
+	/** Wi-Fi Management operations */
+	const struct wifi_mgmt_ops *ops;
```
Shouldn't `struct wifi_mgmt_ops` be rather part of `struct wifi_native_stack`? My understanding is that we want to share a common implementation across all managed interfaces?

krish2718 replied:
Yes, I can move it there.

rlubos commented on subsys/net/l2/wifi/wifi_native.c:
```diff
@@ -0,0 +1,85 @@
+		}
+	}
+
+	mgd_iface = k_malloc(sizeof(*mgd_iface));
+	if (!mgd_iface) {
+		return -ENOMEM;
+	}
```
Do we really need to involve heap here, we usually aim to avoid that in the net stack. I'd prefer to see a preallocated array in `struct wifi_native_stack` instead, especially that such information on what interface will be managed by which supplicant could be easily determined build-time, we can't create network interfaces at runtime in Zephyr.

krish2718 replied:
I will convert linked lists for `stacks` and `mgd_ifaces` to arrays with configuration options.

rlubos reviewed (commented):
I think we need to rethink `wifi_native` nomenclature here. Plus I'd really like to see some documentation (and maybe tests for the new framework), as it bring extra complexity to the Wi-Fi subsystem that may not be very intuitive for newcomers.

krish2718 commented on subsys/net/l2/wifi/wifi_native.c:
```diff
@@ -0,0 +1,85 @@
+
+sys_slist_t wifi_native_stacks = SYS_SLIST_STATIC_INIT(&wifi_native_stacks);
+
+void wifi_native_register_stack(struct wifi_native_stack *stack)
```
Had the same concern here https://github.com/zephyrproject-rtos/zephyr/pull/59249#discussion_r1230814631

krish2718 commented:
> I think we need to rethink `wifi_native` nomenclature here. Plus I'd really like to see some documentation (and maybe tests for the new framework), as it bring extra complexity to the Wi-Fi subsystem that may not be very intuitive for newcomers.

I am writing docs and unit tests already, but raised this as draft to get early feeedback.

krish2718 commented on include/zephyr/net/wifi_mgmt.h:
```diff
@@ -332,6 +320,21 @@ struct net_wifi_mgmt_offload {
+#else
+	struct offloaded_if_api wifi_iface;
+#endif
+	const struct wifi_mgmt_ops *const wifi_api;
```
`wifi_mgmt_api`?

rlubos replied:
Sounds good

krish2718 committed:
tests: wifi: Add Wi-Fi test suite

This adds Wi-Fi tests for both offloaded and newly introduced network
manager managed interface. Only scan op is used to verify the network
manager code.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
doc: net: Add an entry for Wi-Fi Management

This was missing all along.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

krish2718 committed:
doc: net: Add an entry about Wi-Fi Network managers

Add the entry to convey support for Wi-Fi Network Managers in Zephyr.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

rlubos commented on subsys/net/l2/wifi/wifi_mgmt.c:
```diff
@@ -12,6 +12,15 @@ LOG_MODULE_REGISTER(net_wifi_mgmt, CONFIG_NET_L2_WIFI_MGMT_LOG_LEVEL);
 #include <zephyr/net/net_if.h>
 #include <zephyr/net/wifi_mgmt.h>
+#include <zephyr/net/wifi_native.h>
+static const struct wifi_mgmt_ops *const get_wifi_api(struct net_if *iface)
```
Please add an empty line here.

rlubos commented on include/zephyr/net/wifi_nm.h:
```diff
@@ -0,0 +1,100 @@
+ */
+
+/*
+ * Copyright (c) Nordic Semiconductor ASA.
```
Year missing.

rlubos commented on subsys/net/l2/wifi/wifi_nm.c:
```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) Nordic Semiconductor ASA.
```
Year missing.

rlubos commented on include/zephyr/net/wifi_nm.h:
```diff
@@ -0,0 +1,100 @@
+ *
+ * @param name Name of the Network manager instance
+ *
+ * @return struct wifi_nm_instance*  Pointer to Network manager instance
```
There's no need to document returned type, it won't render nicely. Here and below.

rlubos commented on include/zephyr/net/wifi_nm.h:
```diff
@@ -0,0 +1,100 @@
+ * @param nm Pointer to Network manager instance
+ * @param iface Managed interface
+ *
+ * @return int 0 for OK; -EINVAL for invalid parameters; -EEXIST if interface is already registered
```
If you enumerate possible return values, it'd be better to use multiple `@retval` entries.

rlubos commented on subsys/net/l2/wifi/wifi_mgmt.c:
```diff
@@ -28,9 +38,8 @@ static int wifi_connect(uint32_t mgmt_request, struct net_if *iface,
-	struct net_wifi_mgmt_offload *off_api =
-		(struct net_wifi_mgmt_offload *) dev->api;
-	const struct wifi_mgmt_ops *const wifi_mgmt_api = off_api ? off_api->wifi_mgmt_api : NULL;
+
```
No need for this empty line.

rlubos commented on tests/net/wifi/wifi_nm/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+
+cmake_minimum_required(VERSION 3.20.0)
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(wifi)
```
`wifi_nm`

rlubos commented on tests/net/wifi/wifi_nm/src/main.c:
```diff
@@ -0,0 +1,183 @@
+	zassert_equal(nm->ops, &wifi_nm_test_ops,
+		      "Invalid wifi nm ops");
+
+	ret = request_scan();
+	zassert_equal(ret, 0, "Scan request failed");
+	zassert_true(wifi_offload_op_called, "Scan callback not called");
```
Wasn't offloaded case tested in the case above?

krish2718 replied:
Yes, this is offloaded when `nm` is registered but there are no managed interfaces.

rlubos reviewed (commented):
I've left a few more nits, but overall IMO it looks pretty good, the intent of this feature is much clearer now.

krish2718 commented on tests/net/wifi/wifi_nm/src/main.c:
```diff
@@ -0,0 +1,183 @@
+{
+
+	int ret;
+#ifdef CONFIG_WIFI_NM
```
Well, just in case someone wants to disable and check, but I can remove.

rlubos reviewed (dismissed):
LGTM

ssharks commented:
LGTM, I'm not going to give an official approval at this moment as I think it is too important to merge just with 2 green.
